compiler/rustc_mir_transform/src/validate.rs

```diff
@@ -595,6 +595,17 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
         &self,
         pred: impl Upcast<TyCtxt<'tcx>, ty::Predicate<'tcx>>,
     ) -> bool {
+        let pred: ty::Predicate<'tcx> = pred.upcast(self.tcx);
+
+        // We sometimes have to use `defining_opaque_types` for predicates
+        // to succeed here and figuring out how exactly that should work
+        // is annoying. It is harmless enough to just not validate anything
+        // in that case. We still check this after analysis as all opaque
+        // types have been revealed at this point.
+        if pred.has_opaque_types() {
+            return true;
+        }
```

Comment:
I'm slightly concerned because we use `predicate_must_hold_modulo_regions` in several places that don't necessarily have additional later checks. Are there *negative* tests where we can cause compilation failures even when triggering this code-path?

Reply:
Remember that this is just validation I added, and we validate MIR after opaques are revealed. I'm not totally certain if I understand the concern here.